All_materials/20230822_Web_Scraping_Financial_Ombudsman_Service/src/web_scraping_and_data_wrangling/fos_web_scrap.py

```diff
@@ -0,0 +1,335 @@
+import pandas as pd
+from nltk.corpus import stopwords
+import bs4
+import requests
+from io import BytesIO
+import PyPDF2
+from nltk.tokenize import word_tokenize
+from nltk.tokenize import RegexpTokenizer
+import re
+import string
+import datetime
+import gensim
+from nltk.stem import WordNetLemmatizer, SnowballStemmer
+all_stopwords = stopwords.words('english')
+stemmer = SnowballStemmer("english")
+
+#hi
+# Test URL for FOS
+#BASE_URL = "https://www.financial-ombudsman.org.uk/decisions-case-studies/ombudsman-decisions/search?Keyword=scam&Business={banking_code}&DateFrom=2022-01-01&DateTo=2022-06-01&{upheld_code}&Sort=relevance"
+url_link = 'https://www.financial-ombudsman.org.uk/decisions-case-studies/ombudsman-decisions/search?Keyword=scam&IndustrySectorID%5B1%5D=1&DateFrom=2023-01-01&DateTo=2023-06-11&IsUpheld%5B1%5D=1&IsUpheld%5B0%5D=0&Sort=relevance'
+
+
+# You can see that you can use the search within the parameters of the URL
+
+
+def get_info_as_soup(url_link: str):
+    #Getting the request
+    
+    res = requests.get(url_link)
+    
+    #Getting into soup
+    soup = bs4.BeautifulSoup(res.text, "html.parser")
+    
+    return soup
+
+
+def get_list_of_deision(soup):
+    #From the soup to get coorect info
+    
+    li_section = soup.find_all('li')
+    
+    desison_list = []
+
+    for each_li in li_section:
+        
+        text_on_web = each_li.get_text()
+        
+        if 'Decision Reference' in text_on_web:
+            
+            desison_list = desison_list + [text_on_web.split('\n')]
+            
+    return desison_list
+
+def wrangle_each_deision_list(desison_list):
+    
+    list_of_machine_readable_deision = []
+    
+    for each_deison in desison_list:
+        
+        #Wrangle the list
+        each_deison = [re.sub(' +', ' ', x) for x in each_deison if
+                       (x != '')]
+        
+        #Getting the info
+        ref_number = each_deison[0][19:]
+        date_of_desision = datetime.datetime.strptime(each_deison[1], '%d %b %Y')
+        bank = each_deison[2]
+        outcome = each_deison[5].replace(" ", "")
+        descrp = each_deison[8].split(' ', 2)[2][:190]
+        pdf_text = getting_pdf_text(ref_number)
+        
+        dict_each_deision = {
+            'ref':ref_number,
+            'date':date_of_desision,
+            'scale_of_lost': finding_scale_of_money(pdf_text),
+            'bank':bank,
+            'outcome':outcome,
+            'background':descrp,
+            'all_text': pdf_text,
+            'CRM':was_CRM_mentioned(pdf_text)
+            }
+        
+        list_of_machine_readable_deision = list_of_machine_readable_deision + [dict_each_deision]
+        
+    return list_of_machine_readable_deision
+
+def getting_number_of_links_and_soup(soup, url_link):
+    """
+    Getting the list from soup
```

Comment:
what is the list and the soup?
should this be more like- getting a list of links from a chunk of text?

Reply:
Added docstring